### ISSUE TITLE
default_token is always used in splunk_hec_logs sink.

### ISSUE THREAD

**Issue by tsigian** [vectordotdev/vector#16123]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

Is mentioned in the docs for the **splunk_hec_logs** sink that "If an event has a token set in its metadata, it will prevail over the one set here".

![image](https://user-images.githubusercontent.com/5666028/214581800-554dd9e0-6b54-4e34-9a69-f5e7e7b2ca1f.png)

**Our situation:**
We fetch the Splunk token from K8S namespace labels.

**Our action:**
From _'kubernetes_logs'_ source type as input , we do a _'remap'_ transformation and we get the base64 token from a label , then we use the _'set_metadata_field()'_ function to set the _'splunk_hec_token'_ variable in the metadata for the event.

We verified that the 'decode_base64(token)' we use decodes the token correctly.

**Our problem:**
 _'splunk_hec_token'_ metadata variable doesn't exist in the events that vector sends to Splunk.
The behaviour is like the **'set_metadata_field("splunk_hec_token", token)'** code we have runs successfully but it doesn't add anything to the event. 
As a result, the default token is always used and Splunk doesn't accept logs that use a different token.

**Our question:**
How we set the Splunk token in the event's metadata to prevail the **default_token** in the **splunk_hec_logs** sink ?

### Configuration

```text
if .kubernetes.namespace.labels."splunk.logging/forwarding" == "enabled" {
            token = string(.kubernetes.namespace.labels."splunk.logging/token") ?? ""
            if token != "" {
              token, err = decode_base64(token)
              if err == null {
                set_metadata_field("splunk_hec_token", token)
              } else {
                log({
                  "msg":       "Could not decode splunk token",
                  "namespace": .kubernetes.namespace.name,
                  "pod":       .kubernetes.pod.name,
                }, level: "warn", rate_limit_secs: 60)
              }
```

### Version

vector 0.22.3 (x86_64-unknown-linux-musl)

**Comment by davidhuie-dd** [comment]
I just reproduced this, and I can confirm this behavior. I repro'd with this bit of Vector config:

```
[sources.http]
type = "http_server"
address = "0.0.0.0:7000"
encoding = "json"

[sources.http_splunk]
type = "http_server"
address = "0.0.0.0:7001"
path = "/services/collector/event"
encoding = "text"
headers = ["Authorization"]

[transforms.splunk_token]
type = "remap"
inputs = ["http"]
source = '''
%splunk_hec_token = .token
.testing = .token
'''
[sinks.my_sink_id]
type = "splunk_hec_logs"
inputs = [ "splunk_token" ]
endpoint = "http://localhost:7001"
host_key = "hostname"
indexed_fields = [ "field1" ]
compression = "none"
default_token = "asdfasdfasdfasdfasdfasdf"
  [sinks.my_sink_id.encoding]
  codec = "json"
```

You can then run: `curl -v localhost:7000 --data '{"token": "12341234"}'`

In the logs, you don't see that the attached token used in the outgoing HTTP request:
```
{"Authorization":"Splunk asdfasdfasdfasdfasdfasdf","message":"{\"event\":{\"path\":\"/\",\"source_type\":\"http_server\",\"testing\":\"12341234\",\"token\":\"12341234\"},\"fields\":{},\"time\":1674776140.689}","path":"/services/collector/event","source_type":"http_server","timestamp":"2023-01-26T23:35:41.694120131Z"}
```

**Comment by tsigian** [comment]
@davidhuie-dd thank you for taking the time reproducing the issue.

Another attempt was to add the token in the variable **'{{ token }}'** field in a remap transform to be used in the **'default_token'** field as the default token of the **'splunk_hec_logs'** sink for testing purposes, but the variable value doesn't seem to be applied at all in the HTTP request headers and the variable name is used instead of the value. 

In my HTTP trace I can see the authorization header has the variable name instead of the value:
 **'authorization             Splunk {{ token }}'** 
and of course I get **'401 Unauthorized'** from Splunk.

<img width="535" alt="image" src="https://user-images.githubusercontent.com/5666028/215034770-c8fbdb8a-d0f0-42ba-a12b-d523eb162751.png">

Is it possible to use a variable for the **'default_token'** field in **'splunk_hec_logs'** sink ?

**Comment by jszwedko** [comment]
I think the issue here is that you need to use [set_secret](https://vector.dev/docs/reference/vrl/functions/#set_secret) rather than `set_metadata_field`. It's a bit confusing since the docs refer to the token being in the "metadata" but the token is actually pulled from "secrets"

**Comment by jszwedko** [comment]
Hrm, that all looks right to me. I think someone will need to dig into this as a bug.

Vector sends the token as the `Authorization` HTTP header that you found.

**Comment by hargut** [comment]
Tried to reproduce the issue, but for me it works as expected when taking the set_secret into account.

Slightly modified version from the config above:
```
[sources.http]
type = "http_server"
address = "0.0.0.0:7000"
encoding = "json"

[sources.http_splunk]
type = "http_server"
address = "0.0.0.0:7001"
path = "/services/collector/event"
encoding = "text"
headers = ["Authorization"]

[transforms.splunk_token]
type = "remap"
inputs = ["http"]
source = '''
if (.token != null; .token != "")  {
  .token = string!(.token)
  set_secret("splunk_hec_token", .token)
}
'''
[sinks.my_sink_id]
type = "splunk_hec_logs"
inputs = [ "splunk_token" ]
endpoint = "http://localhost:7001"
host_key = "hostname"
indexed_fields = [ "field1" ]
compression = "none"
default_token = "asdfasdfasdfasdfasdfasdf"
[sinks.my_sink_id.encoding]
codec = "json"
```

There is even a test for the pass-through token available at:
https://github.com/vectordotdev/vector/blob/2b28408eae2b3c43a9e88d701e57fc50cd25fad7/src/sinks/splunk_hec/logs/tests.rs#L189

**Comment by tsigian** [comment]
@hargut It works for me also with your config file and 0.27.0 vector version:

**Test setup**:
- Python HTTP server on port 7002 as a fake sink
- Mimtproxy listening on port 7001 and reverse proxing on port 7002 (to python server fake sink)
- Vector 0.27.0 runs with 'set_secret' 

```
$ vector --version
vector 0.27.0 

$ python3 -m http.server 7002

$ mitmproxy --listen-port 7001 --mode reverse:http://localhost:7002

$ cat vector.toml
  [sources.http]
  type = "http_server"
  address = "0.0.0.0:7000"
  encoding = "json"

  [transforms.splunk_token]
  type = "remap"
  inputs = ["http"]
  source = '''
  if (.token != null; .token != "")  {
    .token = string!(.token)
    set_secret("splunk_hec_token", .token)
  }
  '''
  [sinks.my_sink_id]
  type = "splunk_hec_logs"
  inputs = [ "splunk_token" ]
  endpoint = "http://localhost:7001"
  host_key = "hostname"
  indexed_fields = [ "field1" ]
  compression = "none"
  default_token = "asdfasdfasdfasdfasdfasdf"

  [sinks.my_sink_id.encoding]
  codec = "json"

$ vector -c vector.toml
```

**Test 1: Curl without new token, the default token is used:**
```
$ curl -v localhost:7000 --data '{"message": "test1"}'

   HEADER: authorization: Splunk asdfasdfasdfasdfasdfasdf
   BODY:
   {
    "event": {
        "message": "test1",
        "path": "/",
        "source_type": "http_server"
    },
```

**Test 2: Curl with new token, the new token is used:**
```
$ curl -v localhost:7000 --data '{"message": "test2", "token": "test_token-XXXXX"}'

   HEADER: authorization: Splunk test_token-XXXXX
   BODY:
   {
    "event": {
        "message": "test2",
        "path": "/",
        "source_type": "http_server",
        "token": "test_token-XXXXX"
    },
```

**Test 3: Curl with new token, the new token is used:**
```
$ curl -v localhost:7000 --data '{"message": "test3", "token": "test_token-ZZZZZZ"}'

  HEADER: authorization: Splunk test_token-ZZZZZZ
  BODY:
{
    "event": {
        "message": "test3",
        "path": "/",
        "source_type": "http_server",
        "token": "test_token-ZZZZZZ"
    },
```

**Comment by StephenWakely** [comment]
Closed by #16520
